lib/grpc/client/adapters/gun.ex

```diff
@@ -287,4 +321,146 @@ defmodule GRPC.Client.Adapters.Gun do
+        err
+    end
+  end
+
+  defp response_stream(
+         %{
+           channel: %{adapter_payload: ap},
+           response_mod: res_mod,
+           codec: codec,
+           payload: payload
+         },
+         opts
+       ) do
+    state = %{
+      adapter_payload: ap,
+      payload: payload,
+      buffer: <<>>,
+      fin: false,
+      need_more: true,
+      opts: opts,
+      response_mod: res_mod,
+      codec: codec
+    }
+
+    Stream.unfold(state, fn s -> read_stream(s) end)
+  end
+
+  defp read_stream(%{buffer: <<>>, fin: true, fin_resp: nil}), do: nil
+
+  defp read_stream(%{buffer: <<>>, fin: true, fin_resp: fin_resp} = s),
+    do: {fin_resp, Map.put(s, :fin_resp, nil)}
+
+  defp read_stream(
+         %{
+           adapter_payload: ap,
+           payload: payload,
+           buffer: buffer,
+           need_more: true,
+           opts: opts
+         } = s
+       ) do
+    case recv_data_or_trailers(ap, payload, opts) do
+      {:data, data} ->
+        buffer = buffer <> data
+        new_s = s |> Map.put(:need_more, false) |> Map.put(:buffer, buffer)
+        read_stream(new_s)
+
+      {:trailers, trailers} ->
+        trailers = GRPC.Transport.HTTP2.decode_headers(trailers)
+
+        case parse_trailers(trailers) do
+          :ok ->
+            fin_resp =
+              if opts[:return_headers] do
+                {:trailers, trailers}
+              end
+
+            new_s = s |> Map.put(:fin, true) |> Map.put(:fin_resp, fin_resp)
+            read_stream(new_s)
+
+          error ->
+            {error, %{buffer: <<>>, fin: true, fin_resp: nil}}
+        end
+
+      error = {:error, _} ->
+        {error, %{buffer: <<>>, fin: true, fin_resp: nil}}
+    end
+  end
+
+  defp read_stream(%{buffer: buffer, need_more: false, response_mod: res_mod, codec: codec} = s) do
+    case GRPC.Message.get_message(buffer) do
+      # TODO
+      {{_, message}, rest} ->
+        reply = codec.decode(message, res_mod)
+        new_s = Map.put(s, :buffer, rest)
+        {{:ok, reply}, new_s}
+
+      _ ->
+        read_stream(Map.put(s, :need_more, true))
+    end
+  end
+
+  defp parse_response(
+         %{response_mod: res_mod, codec: codec, accepted_compressors: accepted_compressors},
+         headers,
+         body,
+         trailers
+       ) do
+    case parse_trailers(trailers) do
+      :ok ->
+        compressor =
+          case headers do
+            %{"grpc-encoding" => encoding} ->
+              Enum.find(accepted_compressors, nil, fn c -> c.name() == encoding end)
+
+            _ ->
+              nil
+          end
+
+        body =
+          if function_exported?(codec, :unpack_from_channel, 1) do
+            codec.unpack_from_channel(body)
+          else
+            body
+          end
+
+        case GRPC.Message.from_data(%{compressor: compressor}, body) do
+          {:ok, msg} ->
+            {:ok, codec.decode(msg, res_mod)}
+
+          err ->
+            err
+        end
+
```

Comment:
the nesting here seems too much. Perhaps this branch can be a separate defp?

Reply:
yeah! Agreed! I'll refactor this